### PR TITLE
feat(trajectory_follower): extend mpc trajectory for terminal yaw

### PR DESCRIPTION
**Pull request by kosuke55**
## Description

Add extended point to trajectory terminal in yaw direction to arrive on goal pose.

before

https://user-images.githubusercontent.com/39142679/206361614-1d00d6c3-e646-40a4-b9fd-75b760c5118b.mp4

after

https://user-images.githubusercontent.com/39142679/206361589-659cd58c-7ec2-41b6-b14b-6c41e6eea238.mp4

## Related links

[tier4 internal link](https://tier4.atlassian.net/browse/T4PB-23927)

## Tests performed

psim, actual vehicle

tier4 internal scenario tests: https://evaluation.tier4.jp/evaluation/reports/611bfd74-9844-5984-9039-d2c7f69d7999?project_id=prd_jt

821/825 -> 820/825 

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: control/trajectory_follower/src/mpc.cpp [hunk 1]
```diff
@@ -217,6 +218,38 @@ void MPC::setReferenceTrajectory(
     }
   }
 
+  /* extend terminal points */
+  {
```
Comment (TakaHoribe) [inline control/trajectory_follower/src/mpc.cpp:222]: Please make this part a separate function. And the purpose of this function is unclear to other developers. Please write comments.

File: control/trajectory_follower/src/mpc.cpp [hunk 2]
```diff
@@ -217,6 +218,38 @@ void MPC::setReferenceTrajectory(
...
+      const double x_offset = m_is_forward_shift ? extend_interval : -extend_interval;
+      extended_pose = tier4_autoware_utils::calcOffsetPose(extended_pose, x_offset, 0.0, 0.0);
+      const double x = extended_pose.position.x;
+      const double y = extended_pose.position.y;
+      const double z = extended_pose.position.z;
+      // calc relative time
+      const double dx = x - traj.x.back();
+      const double dy = y - traj.y.back();
+      const double dz = z - traj.z.back();
+      const double dist = std::sqrt(dx * dx + dy * dy + dz * dz);
```
Comment (TakaHoribe) [inline control/trajectory_follower/src/mpc.cpp:246]: use hypot.

File: launch/tier4_control_launch/config/trajectory_follower/lateral_controller.param.yaml [hunk 3]
```diff
@@ -40,6 +40,7 @@
     mpc_acceleration_limit: 2.0                  # limit on the vehicle's acceleration
     mpc_velocity_time_constant: 0.3              # time constant used for velocity smoothing
     mpc_min_prediction_length: 5.0               # minimum prediction length
+    mpc_min_vel: 1.0                             # minimum velocity for relative time
```
Comment (TakaHoribe) [inline launch/tier4_control_launch/config/trajectory_follower/lateral_controller.param.yaml:43]: Please check if the control works well for slow-speed products. If not, set this value to 0.1 as previous.

File: control/trajectory_follower/src/mpc.cpp (hunk 1 above)
Comment (kosuke55) [reply control/trajectory_follower/src/mpc.cpp]: fix in 8dc33ca4f5f0fa43396965924f878ddb787151b2

File: control/trajectory_follower/src/mpc.cpp (hunk 2 above)
Comment (kosuke55) [reply control/trajectory_follower/src/mpc.cpp]: fix in 8dc33ca4f5f0fa43396965924f878ddb787151b2

File: launch/tier4_control_launch/config/trajectory_follower/lateral_controller.param.yaml (hunk 3 above)
Comment (kosuke55) [reply launch/tier4_control_launch/config/trajectory_follower/lateral_controller.param.yaml]: remove this param in https://github.com/autowarefoundation/autoware.universe/pull/2447/commits/875a62a1416104f5d2389fac3ca38b4cc889f841, add flag in https://github.com/autowarefoundation/autoware.universe/pull/2447/commits/69e2c0fbf18d55b28add9068c0e1e9e4c2ac65ff.
same to before this PR he min vel in the middle of trajectory is 0.1m/s as a hard code, and 10.0m/s only for exteded points.

File: control/trajectory_follower/src/mpc.cpp [hunk 4]
```diff
@@ -217,6 +217,37 @@ void MPC::setReferenceTrajectory(
     }
   }
 
+  /* extend terminal points.
+   * the original trajectory cannot take into account the pose from the position alone
+   * due to smoothing, etc.
+   */
+  if (enable_traj_extending) {
```
Comment (TakaHoribe) [inline control/trajectory_follower/src/mpc.cpp:224]: Please move thses code into a separate function.

File: control/trajectory_follower/src/mpc_lateral_controller.cpp [hunk 5]
```diff
@@ -65,6 +65,7 @@ MpcLateralController::MpcLateralController(rclcpp::Node & node) : node_{&node}
   m_mpc.m_admissible_yaw_error_rad = node_->declare_parameter<double>("admissible_yaw_error_rad");
   m_mpc.m_use_steer_prediction = node_->declare_parameter<bool>("use_steer_prediction");
   m_mpc.m_param.steer_tau = node_->declare_parameter<double>("vehicle_model_steer_tau");
+  m_enable_traj_extending = node_->declare_parameter<bool>("enable_traj_extending");
```
Comment (TakaHoribe) [inline control/trajectory_follower/src/mpc_lateral_controller.cpp:68]: (naming) the name simply as `enable_traj_extending` is not clear about its purpose. My suggestion is `extend_trajectory_in_yaw_direction`, `extend_trajectory_for_end_yaw_control` or something like developpers can understand its purpose.

File: control/trajectory_follower/src/mpc.cpp (hunk 1 above)
Comment (TakaHoribe) [reply control/trajectory_follower/src/mpc.cpp]: Seems not addressed in https://github.com/autowarefoundation/autoware.universe/commit/8dc33ca4f5f0fa43396965924f878ddb787151b2 ?

File: control/trajectory_follower/src/mpc.cpp [hunk 6]
```diff
@@ -217,6 +217,37 @@ void MPC::setReferenceTrajectory(
...
+    constexpr double extend_vel = 10.0;
+    const double extend_interval = traj_resample_dist;
+    const size_t num_extended_point = static_cast<size_t>(extend_dist / extend_interval);
+    for (size_t i = 0; i < num_extended_point; ++i) {
+      const double x_offset = m_is_forward_shift ? extend_interval : -extend_interval;
+      extended_pose = tier4_autoware_utils::calcOffsetPose(extended_pose, x_offset, 0.0, 0.0);
+      const double x = extended_pose.position.x;
+      const double y = extended_pose.position.y;
+      const double z = extended_pose.position.z;
+      const double dist = std::hypot(x - traj.x.back(), y - traj.y.back(), z - traj.z.back());
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply control/trajectory_follower/src/mpc.cpp]: thanks 
`dist = extend_interval`

[the remark this answers is not in the document]
Comment (kosuke55) [reply control/trajectory_follower/src/mpc.cpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/2447/commits/c0cf295e6c81d5fd4382415f3096a18da30f9ac5

File: control/trajectory_follower/src/mpc.cpp (hunk 4 above)
Comment (kosuke55) [reply control/trajectory_follower/src/mpc.cpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/2447/commits/c0cf295e6c81d5fd4382415f3096a18da30f9ac5

File: control/trajectory_follower/src/mpc.cpp [hunk 7]
```diff
@@ -217,6 +217,37 @@ void MPC::setReferenceTrajectory(
     }
   }
 
+  /* extend terminal points.
+   * the original trajectory cannot take into account the pose from the position alone
+   * due to smoothing, etc.
```
[the remark this answers is not in the document]
Comment (kosuke55) [reply control/trajectory_follower/src/mpc.cpp]: fixed in https://github.com/autowarefoundation/autoware.universe/pull/2447/commits/c0cf295e6c81d5fd4382415f3096a18da30f9ac5

File: control/trajectory_follower/src/mpc_lateral_controller.cpp (hunk 5 above)
Comment (kosuke55) [reply control/trajectory_follower/src/mpc_lateral_controller.cpp]: I chose `extend_trajectory_for_end_yaw_control`
fixed in https://github.com/autowarefoundation/autoware.universe/pull/2447/commits/c0cf295e6c81d5fd4382415f3096a18da30f9ac5

**Review by TakaHoribe — APPROVED** [review]
LGTM

**Pull request merged**